### DISCUSSION TITLE
Catalog Fetching

### DISCUSSION THREAD

**Discussion by Akshay-a-j** [eclipse-edc/Samples#77] [Q&A]
I was trying to test permission in transfer-06 example where I made few changes in step 4. Create a Policy on the provider as shown below. 
```
curl -d '{
"@context": {
"edc": "https://w3id.org/edc/v0.0.1/ns/",
"odrl": "http://www.w3.org/ns/odrl/2/"
},
"@id": "aPolicy",
"policy": {
"@type": "set",
"odrl:permission": [
        {
            "edctype": "dataspaceconnector:permission",
            "uid": null,
            "target": "assetId",
            "action": {
                "type": "USE",
                "includedIn": null,
                "constraint": null
            },
            "assignee": "assignee",
            "assigner": "assigner",
            "constraints": [
                {
                    "edctype": "AtomicConstraint",
                    "leftExpression": {
                        "edctype": "dataspaceconnector:literalexpression",
                        "value": "ids:absoluteSpatialPosition"
                    },
                    "rightExpression": {
                        "edctype": "dataspaceconnector:literalexpression",
                        "value": "eu"
                    },
                    "operator": "EQ"
                }
            ],
            "duties": []
        }
    ],
"odrl:prohibition": [],
"odrl:obligation": []
}
}' -H 'content-type: application/json' http://localhost:19193/management/v2/policydefinitions \
-s | jq
```

Now, upon executing step 6. How to fetch catalog on consumer side, i.e.,
```
curl -X POST "http://localhost:29193/management/v2/catalog/request" \
    -H 'Content-Type: application/json' \
    -d '{
      "@context": {
        "edc": "https://w3id.org/edc/v0.0.1/ns/"
      },
      "providerUrl": "http://localhost:19194/protocol",
      "protocol": "dataspace-protocol-http"
    }' -s | jq
```
it throws following error: 
```{
    "message": "org.eclipse.edc.spi.http.EdcHttpClientException: Server response to Request{method=POST, url=http://localhost:19194/protocol/catalog/request, headers=[Content-Type:application/json, Authorization:{\"region\":\"eu\",\"audience\":\"http://localhost:19194/protocol\",\"clientId\":\"consumer\"}]} was not successful but was 
500: <html>\n<head>\n<meta http-equiv=\"Content-Type\" content=\"text/html;charset=ISO-8859-1\"/>\n<title>Error 
500 Internal Server Error</title>\n</head>\n<body><h2>HTTP ERROR 500 Internal Server Error</h2>\n<table>\n<tr><th>URI:</th><td>/protocol/catalog/request</td></tr>\n<tr><th>STATUS:</th><td>500</td></tr>\n<tr><th>MESSAGE:</th><td>Internal Server Error</td></tr>\n<tr><th>SERVLET:</th><td>EDC-protocol</td></tr>\n</table>\n<hr/><a href=\"https://eclipse.org/jetty\">Powered by Jetty:// 11.0.15</a><hr/>\n\n</body>\n</html>\n",
    "type": "BadGateway",
    "path": null,
    "invalidValue": null
  }
```
Can someone tell me why exactly it is throwing the abov error?
Thank you

**Comment by Akshay-a-j** [comment]
At the provider connector end, it throws following error:
SEVERE 2023-07-05T11:04:14.7568949 JerseyExtension: Unexpected exception caught
java.lang.NullPointerException: Cannot invoke "org.eclipse.edc.policy.model.Action.getType()" because "action" is null

**Reply by ndr-brt** [reply]
looking at the policy you are using, seems you are using the old EDC format, while the Policy should follow the ODRL ontology.
Likely this should have been blocked by a validation, that's not in place yet.

We are also working on the documentation, in the meantime for more info you could take a look at this document in a PR on an EDC user project (tractusx): https://github.com/eclipse-tractusx/tractusx-edc/pull/441/files#diff-8f2b4c82df965b579abd5f4b6f7917fa12ae4350e1756bf118dbcacae2a4616a

**Comment by Akshay-a-j** [comment]
Thank you @ndr-brt. Its working now. However, I hav eone more question regrading the policy:permission. The above set of code tells us about the policy on provider side. Now if I want to initiate contract negotiation from consumer connector, what would be my leftoperand and rightoperand?

NOTE: <This is how I am defining permission on provider connector:
 "odrl:permission": [
      {
        "odrl:action": "USE",
        "odrl:constraint": [
          {
            "odrl:leftOperand": {"@value":"ids:absoluteSpatialPosition"},
            "odrl:operator": "EQ",
            "odrl:rightOperand":  {"@value":"eu"}
          }]
      }
    ],>

Thank you

**Reply by ndr-brt** [reply]
in the standard flow, the consumer would just take the policy from the catalog and put it in the negotiation request, because currently there's a "policy equality" validator: if the policy in the negotiation is not the same that was in the contract offer, the negotiation will fail.

**Comment by Akshay-a-j** [comment]
You mean to say the permission section in [policy](https://github.com/eclipse-edc/Samples/tree/main/transfer/transfer-07-provider-push-http#3-create-a-policy-on-the-provider) and [contract negotiation](https://github.com/eclipse-edc/Samples/tree/main/transfer/transfer-07-provider-push-http#6-negotiate-a-contract) should be same? 
If so, I tried with same permission mentioned above, still it is failed to validate. 
![Screenshot (70)](https://github.com/eclipse-edc/Samples/assets/38350400/6b7e0ef4-7296-4b89-abe9-429ce5cb9bd9)

**Reply by ndr-brt** [reply]
no, it should be the same as the one in the catalog's `dataset`.
which kind of validation error do you receive?

**Reply by Akshay-a-j** [reply]
Thank you @ndr-brt. Its working now.

**Comment by Akshay-a-j** [comment]
I am currently exploring the implementation of Prohibition and Obligation methods within policies. Could you kindly provide some reference links to existing examples where these methods have been implemented in EDC?

**Reply by ndr-brt** [reply]
The EDC itself does not provide a policy enforcement model, the only policy function that's given out of the box is the "expiration" one:
https://github.com/eclipse-edc/Connector/blob/034258ea8b135330cb59c039d9d06aca1079db96/core/control-plane/contract-core/src/main/java/org/eclipse/edc/connector/contract/validation/ContractExpiryCheckFunction.java

you can get some examples on how it can be used in the Connector end to end tests: https://github.com/eclipse-edc/Connector/blob/main/system-tests/e2e-test-fixtures/src/testFixtures/java/org/eclipse/edc/test/system/utils/PolicyFixtures.java#L60
